st2actions/st2actions/runners/ssh/parallel_ssh.py

```diff
@@ -0,0 +1,317 @@
+        try:
+            LOG.debug('Running command: %s on host: %s.', cmd, host)
+            client = self._hosts_client[host]
+            (stdout, stderr, exit_code) = client.run(cmd, timeout=timeout)
+            is_succeeded = (exit_code == 0)
+            results[host] = {'stdout': stdout, 'stderr': stderr, 'return_code': exit_code,
+                             'succeeded': is_succeeded, 'failed': not is_succeeded}
+        except:
+            error = 'Failed executing command %s on host %s', cmd, host
+            LOG.exception(error)
+            _, _, tb = sys.exc_info()
+            results[host] = self._generate_error_result(error, tb)
+
+    def _put_files(self, local_path, remote_path, host, results, mode=None,
+                   mirror_local_mode=False):
+        try:
+            LOG.debug('Copying file to host: %s' % host)
+            if os.path.isdir(local_path):
+                result = self._hosts_client[host].put_dir(local_path, remote_path)
+            else:
+                result = self._hosts_client[host].put(local_path, remote_path,
+                                                      mirror_local_mode=mirror_local_mode,
+                                                      mode=mode)
+            LOG.debug('Result of copy: %s' % result)
+            results[host] = result
+        except:
+            error = 'Failed sending file(s) in path %s to host %s' % (local_path, host)
+            LOG.exception(error)
+            _, _, tb = sys.exc_info()
+            results[host] = self._generate_error_result(error, tb)
+
+    def _mkdir(self, host, path, results):
+        try:
+            result = self._hosts_client[host].mkdir(path)
+            results[host] = result
+        except:
+            error = 'Failed "mkdir %s" on host %s.' % (path, host)
+            LOG.exception(error)
+            _, _, tb = sys.exc_info()
+            results[host] = self._generate_error_result(error, tb)
+
+    def _delete_file(self, host, path, results):
+        try:
+            result = self._hosts_client[host].delete_file(path)
+            results[host] = result
+        except:
+            error = 'Failed deleting file %s on host %s.' % (path, host)
+            LOG.exception(error)
+            _, _, tb = sys.exc_info()
+            results[host] = self._generate_error_result(error, tb)
+
+    def _delete_dir(self, host, path, results, force=False, timeout=None):
+        try:
+            result = self._hosts_client[host].delete_dir(path, force=force, timeout=timeout)
+            results[host] = result
+        except:
+            error = 'Failed deleting dir %s on host %s.' % (path, host)
+            LOG.exception(error)
+            _, _, tb = sys.exc_info()
+            results[host] = self._generate_error_result(error, tb)
+
+    def _get_host_port_info(self, host_str):
+        hostname = host_str
+        port = self._ssh_port
+        if ':' in hostname:
```

Comment:
Looks like latest version of `eventlet` now includes IPv6 support so it would also make sense to make this code a bit more robust so it can also handle IPv6 addresses :P